### PR TITLE
add darkside Reset, StageBlocks*, ApplyStaged, GetIncomingTransactions for reorg testing

### PR DESCRIPTION
**Pull request by LarryRuane**
Addresses #242, #235, #199.

### CODE REVIEW

File: common/darkside.go [hunk 1]
```diff
@@ -5,49 +5,176 @@ import (
...
+	saplingActivation int
+	branchID          string
+	chainName         string
 	// Should always be nonempty. Index 0 is the block at height start_height.
-	blocks                []string
-	incoming_transactions [][]byte
-	server_start          time.Time
+	blocks               [][]byte // full blocks, binary, as from zcashd getblock rpc
+	incomingTransactions [][]byte // full transactions, binary, zcashd getrawtransaction txid
+	serverStart          time.Time
```
Comment (LarryRuane) [inline common/darkside.go:27]: I started using vscode, and it has a built-in linter, so some of the changes here are just renaming it suggested, and other formatting with no functional effect.

File: docs/darksidewalletd.md [hunk 2]
```diff
@@ -93,20 +93,29 @@ submit the blocks, which internally uses grpcurl, e.g.:
...
+
+If you submit overlapping ranges, the expected things happen. For example, first
+submit 1000-1005, then 1003-1007, the result is 1000-1007 (the original 1000-1002
+followed by the new 1003-1007). This is how you can create a reorg starting at 1003.
+You can get the same effect slightly less efficiently by submitting 1000-1007 (that
+is, resubmitting the original 1000-1002 followed by the new 1003-1007).
+
+If you first submit 1000-1005, then 1001-1002, the result will be 1000-1002
+(1003-1005 are dropped; it's not possible to "insert" blocks into an range).
+Likewise, first submit 1005-1008, then 1000-1006, the result is only 1000-1006. As
```
Comment (defuse) [inline docs/darksidewalletd.md:115]:
```suggestion
Likewise, first submit 1005-1008, then 1000-1006, the result is only 1000-1006. An
```

File: common/darkside.go [hunk 3]
```diff
@@ -5,49 +5,176 @@ import (
...
+
+func DarksideSetBlocksURL(url string) error {
+	if strings.HasPrefix(url, "file:") && len(url) >= 6 && url[5] != '/' {
+		dir, err := os.Getwd()
+		if err != nil {
+			return err
+		}
+		url = "file:" + dir + string(os.PathSeparator) + url[5:]
+	}
+	cmd := exec.Command("curl", "--silent", "--show-error", url)
```
Comment (defuse) [inline common/darkside.go:146]: This has a couple security implications, which I think are fine for test code but they should be documented *very visibly*:

- With `file:` URLs path traversal is possible, i.e. someone could submit the URL `file:foo/../../../../../../../../etc/passwd` and this code will try to read the system's `/etc/passwd` file as blocks. This will probably fail and be useless to the attacker, but it's *possible* for there to be hex-encoded files containing keys or something on the developer's system that could be leaked this way. (It's possible to avoid path traversal with careful sanitation of the paths. Consider using something like: https://github.com/cyphar/filepath-securejoin)
- Being able to get someone's machine to make arbitrary web requests could be slightly dangerous, e.g. make them request `https://nasa.gov/some/url/that/will/hack/nasa?exploit=payloadbytes`.

Suggested doc text:

> Leaving darksidewalletd puts your machine at greater risk because (a) it may be possible to use `file:` paths with `DarksideSetBlocksURL` to read arbitrary files on your system, and (b) also using `DarksideSetBlocksURL`, someone can force your system to make a web request to an arbitrary URL (which could have your system download questionable material, perform attacks on other systems, etc.). The maximum 30-minute run time limit built into darksidewalletd mitigates these risks, but users should still be cautious.

File: common/darkside.go [hunk 4]
```diff
@@ -5,49 +5,176 @@ import (
...
+		return errors.New(string(stderroutstr))
+	}
+	return nil
+}
+
+func DarksideSendTransaction(txbytes []byte) {
+	state.incomingTransactions = append(state.incomingTransactions, txbytes)
+}
+func darksideRawRequest(method string, params []json.RawMessage) (json.RawMessage, error) {
+	if time.Now().Sub(state.serverStart).Minutes() >= 30 {
```
Comment (defuse) [inline common/darkside.go:177]: At at glance at the code, it isn't obvious to me that this check happens for *every* darkside operation, e.g. I think `DarksideSetMetaState` or `DarksideSetBlocksURL` could be executed without triggering the check. Let's extract this into a helper method called `darksideShutdownIfRunningTooLong` (or similar) and call it at the start of all the public functions in this file so it's really obvious the check always happens.

File: common/darkside.go [hunk 5]
```diff
@@ -5,49 +5,176 @@ import (
...
+			return err
 		}
-		scan := bufio.NewScanner(testBlocks)
-		for scan.Scan() { // each line (block)
-			block := scan.Bytes()
-			state.blocks = append(state.blocks, string(block))
+		if len(rest) != 0 {
+			return errors.New("block is too long")
 		}
+		copy(blockData[4:4+32], prevblock.GetEncodableHash())
```
Comment (defuse) [inline common/darkside.go:101]: I'm thinking that we shouldn't automatically patch up the `prevHash` here. By not doing so, testers retain full control over the blocks' prevHashes (aside from constraints on them set by how lightwalletd handles reorgs). That might be useful for testing lightwalletd or wallets for very-misbehaving zcashd behavior. Also by not doing it here, it would make it obvious when our dataset blocks checked into the other repo have invalid prevHashes.

File: testtools/genblocks/main.go [hunk 6]
```diff
@@ -1,7 +1,25 @@
...
+// transaction and all of the transactions in the corresponding file.
+
+// The default start height is 1000, so the program expects to find
+// files blocks/1000.txt, blocks/1001.txt, ...
+//
+// Typical way to run this program to create 6 blocks, all empty except
+// for the fifth, which contains one transaction:
 //     $ mkdir blocks
 //     $ touch blocks/{1000,1001,1002,1003,1004,1005}.txt
+//     $ echo "0400008085202f8901950521a79e89ed418a4b506f42e9829739b1ca516d4c590bddb4465b4b347bb2000000006a4730440220142920f2a9240c5c64406668c9a16d223bd01db33a773beada7f9c9b930cf02b0220171cbee9232f9c5684eb918db70918e701b86813732871e1bec6fbfb38194f53012102975c020dd223263d2a9bfff2fa6004df4c07db9f01c531967546ef941e2fcfbffeffffff026daf9b00000000001976a91461af073e7679f06677c83aa48f205e4b98feb8d188ac61760356100000001976a91406f6b9a7e1525ee12fd77af9b94a54179785011b88ac4c880b007f880b000000000000000000000000" > blocks/1004.txt
```
Comment (LarryRuane) [inline testtools/genblocks/main.go:18]: I dislike this long line, but I really wanted to capture this transaction hex here, which I got from your ticket https://github.com/zcash/lightwalletd/issues/204, since it's very useful.

File: docs/darksidewalletd.md [hunk 7]
```diff
@@ -143,7 +152,7 @@ Now you can start darksidewalletd and it’ll load the blocksA blocks:
 That will have loaded and be serving the blocksA blocks. We can push up the
 blocksB blocks using ./utils/submitblocks.sh:
 
-`./utils/submitblocks.sh 1000 1000 testdata/darkside-blocks-reorg`
+`./utils/submitblocks.sh 1000 testdata/darkside-blocks-reorg`
```
[the remark this answers is not in the document]
Comment (defuse) [reply docs/darksidewalletd.md:155]: Yep that should be fine :)

File: common/darkside.go [hunk 8]
```diff
@@ -5,49 +5,176 @@ import (
...
+	var scanbuf []byte
+	scan.Buffer(scanbuf, 8*1000*1000)
+	for scan.Scan() { // each line (block)
+		if err := DarksideAddBlock(scan.Text()); err != nil {
+			return err
+		}
+	}
+	if scan.Err() != nil {
+		return scan.Err()
+	}
```
[the remark this answers is not in the document]
Comment (defuse) [reply common/darkside.go:123]: Wow! That's poor design on go's part!

File: frontend/service.go [hunk 9]
```diff
@@ -221,6 +220,24 @@ func (s *LwdStreamer) SendTransaction(ctx context.Context, rawtx *walletrpc.RawT
 	// Result:
 	// "hex"             (string) The transaction hash in hex
 
+	if common.DarksideIsEnabled() {
```
Comment (gmale) [inline frontend/service.go:223]: It seems like this function should possibly be separate. Should we consider removing this `if common.DarksideIsEnabled()` check and creating a separate function, instead?

File: common/darkside.go (hunk 5 above)
Comment (LarryRuane) [reply common/darkside.go:101]: I see what you mean, but somehow I'm not convinced. Yes, it's good to throw bad input at the code under test to see how it reacts, which is the whole idea of fuzzing. But this isn't bad input. This can happen in real life with a zcashd that isn't misbehaving. The lightwalletd can only interpret an incorrect prevhash as a reorg, which is already being tested in various ways. The difference would be that this would be a _persistent_ incorrect prevhash, but that's not qualitatively different -- it's theoretically _possible_ for 100 one-block reorgs to occur in a row. 

After 100 consecutive reorgs (from lightwalletd's point of view), which occurs automatically by zcashd insisting on presenting a block with the wrong prevhash long enough, lightwalletd will crash. (I'm not sure it should, by the way, but that's a separate topic.) Do we want tests that cause lightwalletd to crash? There are no such tests so far, but if we do want such a test, how will we detect lightwalletd crashed, and crashed for the right reason? Open and inspect the log file? Also, how long do we wait? The lightwalletd sleeps a few seconds between attempts to resolve a reorg, which makes the timing to reach 100 reorgs somewhat random. (Such a test will take a few hundred seconds, which is undesirable.)

I added this prevhash patching to make it more convenient for the tests so they wouldn't have to always submit blocks from the beginning. That requirement seems inefficient and brittle to me. I'd like to move to the `Generate()` gRPC that would create a single block at a time. (Speaking of which, the zcash Python rpc-tests have a `generate` test RPC, which inspired my suggestion for `Generate()`, and that doesn't provide a way to create an incorrect prevhash. Has that caused a testing deficiency? That's an admittedly weak point, but thought I'd mention it.)

File: common/darkside.go (hunk 4 above)
Comment (defuse) [reply common/darkside.go]: Nice!

File: common/darkside.go (hunk 5 above)
Comment (defuse) [reply common/darkside.go:106]: I agree about how little use it would be for testing. The only case we're missing out on is a case where zcashd is persistently presenting data with a broken `prevHash` which should never happen.

Another way to say what I meant is: If this bit of code actually fixes up the `prevHash`, it means the set of blocks we loaded into darksidewalletd has a broken `prevHash` somewhere. Whoever made that dataset probably wants to find out quickly instead of having their mistake hidden from them (since it's a sign there could be *other* problems with the dataset). For `Generate()`, its implementation can bare the burden of using the correct `prevHash` before passing the block to this function.

Maybe a compromise is to emit log messages if the prevHash was not actually correct?

By the way, you might be interested in the background that leads me to have the opposite intuition: http://langsec.org/papers/postel-patch.pdf. tl;dr: There are two competing design philosophies, one is "handle unexpected errors gracefully" and the other is "crash and make a lot of noise when anything unexpected happens". It's harder to get things up and running by following the latter principle, but it has several advantages: code ends up being simpler, it's easier to test because there aren't as many edge cases, and bugs in Component A aren't hidden by Component B's graceful handling of Component A's misbehavior.

File: common/darkside.go (hunk 3 above)
Comment (defuse) [reply common/darkside.go:151]: I want to retain the ability to submit arbitrary blocks so that once the wallets implement block header validation, we can write tests for that. (Once wallets start validating headers, `Generate()` will presumably always emit correct headers for the blocks it creates).

File: frontend/service.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (LarryRuane) [reply frontend/service.go:226]: Fixed by a2e88ac54fcd0072e0d8d7846b3281e02c4f7d49 or at least I think that's what you're suggesting.
